### ISSUE TITLE
Dask-friendly 1D interpolation

### ISSUE THREAD

**Issue by crusaderky** [scipy/scipy#8810]
I'm writing an xarray+dask library to perform 1D interpolation (https://github.com/crusaderky/xarray_extras/blob/master/xarray_extras/interpolate.py).
The library takes a ND xarray with dask backend as y and one of its coords (a plain numpy array) as x.
Currently I'm wrapping around ``scipy.interpolate.make_interp_spline``, and it's not great for a couple of reasons:

1. when working in dask, I might have multiple chunks on any axis of y except the interpolation axis; each chunk will translate to an equally-shaped array of spline coefficients. However, across all chunks there is only one array of knots, which (since my x is always a pure numpy array) is also a pure numpy array. With ``make_interp_spline``, I am forced to needlessly re-compute the knots vector multiple times.
2. dask makes it super easy to wrap around functions that return numpy arrays. ``make_interp_spline`` however returns a ``BSpline`` object, from which I have to extract the ``t`` and ``c`` arrays.

My current code looks pretty awful:
```
spline = map_blocks(
    make_interp_spline,
    x, a.data, k=k, check_finite=False, dtype=float)
t_name = 'spline_t-' + tokenize(getattr, spline.name, 't')
spline_first_key = (spline.name, ) + (0, ) * spline.ndim
t = Array(
    {(t_name, 0): (getattr, spline_first_key, 't')},
    name=t_name, dtype=float,
    chunks=((x.size + k + 1, ), ))
    t.dask.update(spline.dask)
c = map_blocks(getattr, spline, 'c', dtype=float)
```
Besides looking ugly and needlessly recomputing t for each chunk, it also introduces a dependency on the first chunk from all other chunks in order to save memory (the spline evaluation of all chunks will read t from the first chunk). It would have been much nicer to just compute t in plain numpy.

# Proposed design
Gut ``make_interp_spline`` and create two new functions out of it, ``make_interp_knots(x, k)`` and ``make_interp_coeffs(y, t, k)``, both of which return  a plain np.ndarray.
``make_interp_spline`` would become a simple wrapper around those two. The only non-trivial piece of logic it would retain is the special short-circuit for ``k=0`` and ``(k1=1 and t=None)`` (https://github.com/scipy/scipy/blob/v1.0.0/scipy/interpolate/_bsplines.py#L706).

**Comment by ev-br** [comment]
1. Can you construct the knot vector once (copy-paste private finctions _not_a_knot/_augknt, both are trivial) and feed it to make_interp_spline(..., t=t)?

2. We had a discussion of whether make_interp_spline should return a tck-tuple or a BSpline object, in the PR which introduced it.

**Comment by ev-br** [comment]
Am circling back to this now. What we can do is to refactor `make_interp_spline` into a wrapper over two leading-underscore functions, roughly `_make_interp_knots(x, k, bc_type)` and `_make_interp_coefs(x, y, t, k, bc_type)`. These two functions will definitely *not* be public, no technically there will be no guarantees about them. However it's likely that they be quite stable on a timescale of several years at least. Then it's up to xarray or xarray_extras to decide if they want to reach into it and use these functions. Would that help @crusaderky ?

**Comment by crusaderky** [comment]
I haven't looked at this problem for 4 years, so my memory is failing me.
Private functions are better than having to reimplement everything from zero.
I think they're missing the `axis` and `check_finite` arguments (see my implementation above).

**Comment by ev-br** [comment]
Indeed, these arguments are missing in my rough sketch above, should be added where needed in the real implementation. So if it's helpful for dask, PR is welcome (will review in less than four years, hopefully).

**Comment by crusaderky** [comment]
I'm afraid I stopped working on this a long time ago; xarray_extras as a whole is in long term maintenance only from my side.
Still I'll be happy to update xarray_extrras to use the scipy implementation if someone else codes it in.
